File: 3_Longest_Substring_Without_Repeating_Characters/3_Longest_Substring_Without_Repeating_Characters.py
__author__ = "CodeamonCat"
__email__ = "CodeamonCat@gmail.com"


class Solution:

    def lengthOfLongestSubstring(self, s: str) -> int:
        left, max_s = 0, 0
        last_index = dict()

        for right, value in enumerate(s):
            left = max(left, last_index[value] +
                       1) if value in last_index else left
            max_s = max(max_s, right - left + 1)
            last_index[value] = right

        return max_s

File: 3_Longest_Substring_Without_Repeating_Characters/3_Longest_Substring_Without_Repeating_Characters.java
/*
 * @author CodeamonCat/CodeamonCat@gmail.com
 */

class Solution {
  public int lengthOfLongestSubstring(String s) {
    int max_s = 0;
    HashMap<Character, Integer> hashtable = new HashMap<>();

    for (int left = 0, right = 0; right < s.length(); right++) {
      if (hashtable.containsKey(s.charAt(right))) {
        left = Math.max(left, hashtable.get(s.charAt(right)));
      }
      max_s = Math.max(max_s, right - left + 1);
      hashtable.put(s.charAt(right), right + 1);
    }

    return max_s;
  }
}
